akd/src/proto/mod.rs

```diff
@@ -0,0 +1,287 @@
+        let len = std::cmp::min(raw.len(), 32);
+        // construct the output buffer
+        let mut out_val = [0u8; 32];
+        // copy into the output buffer the raw data up to the computed length
+        out_val[..len].clone_from_slice(&raw[..len]);
+
+        Ok(crate::NodeLabel {
+            label_len: input.get_label_len(),
+            label_val: out_val,
+        })
+    }
+}
+
+// ==============================================================
+// Node
+// ==============================================================
+
+impl From<&crate::helper_structs::Node<Hasher>> for audit::Node {
+    fn from(input: &crate::helper_structs::Node<Hasher>) -> Self {
+        let mut result = Self::new();
+        result.set_label((&input.label).into());
+        result.set_hash(hash_to_bytes!(input.hash));
+        result
+    }
+}
+
+impl TryFrom<&audit::Node> for crate::helper_structs::Node<Hasher> {
+    type Error = String;
+
+    fn try_from(input: &audit::Node) -> Result<Self, Self::Error> {
+        require!(input, has_label);
+        require!(input, has_hash);
+        let label: crate::NodeLabel = input.get_label().try_into()?;
+        Ok(crate::helper_structs::Node::<Hasher> {
+            label,
+            hash: hash_from_bytes!(input.get_hash()),
+        })
+    }
+}
+
+impl From<&crate::proof_structs::SingleAppendOnlyProof<Hasher>> for audit::SingleEncodedProof {
+    fn from(input: &crate::proof_structs::SingleAppendOnlyProof<Hasher>) -> Self {
+        let mut result = Self::new();
+        let inserted = input
+            .inserted
+            .iter()
+            .map(|item| item.into())
+            .collect::<Vec<_>>();
+        let unchanged = input
+            .unchanged_nodes
+            .iter()
+            .map(|item| item.into())
+            .collect::<Vec<_>>();
+
+        result.set_inserted(RepeatedField::from_vec(inserted));
+        result.set_unchanged(RepeatedField::from_vec(unchanged));
+        result
+    }
+}
+
+impl TryFrom<audit::SingleEncodedProof> for crate::proof_structs::SingleAppendOnlyProof<Hasher> {
+    type Error = String;
+
+    fn try_from(input: audit::SingleEncodedProof) -> Result<Self, Self::Error> {
+        let mut inserted = vec![];
+        let mut unchanged = vec![];
+        for item in input.get_inserted() {
+            inserted.push(item.try_into()?);
+        }
+        for item in input.get_unchanged() {
+            unchanged.push(item.try_into()?);
+        }
+        Ok(crate::proof_structs::SingleAppendOnlyProof {
+            inserted,
+            unchanged_nodes: unchanged,
+        })
+    }
+}
+
+// ************************ Helper Functions ************************ //
+
+const NAME_SEPARATOR: char = '/';
+
+/// The constructed blobs with naming encoding the
+/// blob name = "EPOCH/PREVIOUS_ROOT_HASH/CURRENT_ROOT_HASH"
+pub struct AuditBlob {
+    /// The name of the blob, which can be decomposed into logical components (phash, chash, epoch)
+    pub name: String,
```

Comment:
This feels like a good spot to leverage something like the [new type idiom](https://doc.rust-lang.org/rust-by-example/generics/new_types.html) to denote what exactly composes a blob name instead of simply relying on a comment. We'd be potentially be guarded against funky ordering as well, as the comments already specify a different order for previous hash, current hash, and epoch based on ordering on line 159 vs 161.